File: android/measure-android/measure/src/main/java/sh/measure/android/logs/LogEventCollector.kt
package sh.measure.android.logs

import sh.measure.android.attributes.AttributeValue
import sh.measure.android.config.ConfigProvider
import sh.measure.android.events.EventType
import sh.measure.android.events.SignalProcessor
import sh.measure.android.logger.LogLevel
import sh.measure.android.logger.Logger
import sh.measure.android.utils.TimeProvider
import java.util.concurrent.atomic.AtomicBoolean

internal class LogEventCollector(
    private val logger: Logger,
    private val signalProcessor: SignalProcessor,
    private val timeProvider: TimeProvider,
    private val configProvider: ConfigProvider,
) {
    private val isEnabled = AtomicBoolean(false)

    fun register() {
        isEnabled.set(true)
    }

    fun unregister() {
        isEnabled.set(false)
    }

    val isAutomaticCollectionEnabled: Boolean
        get() = configProvider.logAutocollectEnabled

    fun trackLog(
        body: String,
        severity: LogSeverity,
        attributes: Map<String, AttributeValue> = emptyMap(),
        userTriggered: Boolean = true,
    ) {
        if (!isEnabled.get()) {
            return
        }
        if (body.isEmpty()) {
            logger.log(LogLevel.Error, "Invalid log: body is empty")
            return
        }
        if (severity.severityNumber < configProvider.logMinSeverity) {
            return
        }
        if (configProvider.shouldDiscardLog(body)) {
            return
        }
        val data = LogData(
            severityText = severity.value,
            severityNumber = severity.severityNumber,
            body = body.take(configProvider.maxLogBodyLength),
        )
        signalProcessor.track(
            timestamp = timeProvider.now(),
            type = EventType.LOG,
            data = data,
            userTriggered = userTriggered,
            userDefinedAttributes = attributes,
        )
    }
}

File: ios/Sources/MeasureSDK/Swift/Events/LogEventCollector.swift
//
//  LogEventCollector.swift
//  Measure
//
//  Created by Abhay Sood on 11/06/26.
//

import Foundation

protocol LogEventCollector {
    func enable()
    func disable()
    func trackLog(body: String, severity: LogSeverity, attributes: [String: AttributeValue])
}

final class BaseLogEventCollector: LogEventCollector {
    private let logger: Logger
    private let signalProcessor: SignalProcessor
    private let timeProvider: TimeProvider
    private let configProvider: ConfigProvider
    private let attributeValueValidator: AttributeValueValidator
    private var isEnabled = AtomicBool(false)

    init(logger: Logger, signalProcessor: SignalProcessor, timeProvider: TimeProvider, configProvider: ConfigProvider, attributeValueValidator: AttributeValueValidator) {
        self.logger = logger
        self.signalProcessor = signalProcessor
        self.timeProvider = timeProvider
        self.configProvider = configProvider
        self.attributeValueValidator = attributeValueValidator
    }

    func enable() {
        isEnabled.setTrueIfFalse {
            logger.log(level: .info, message: "LogEventCollector enabled.", error: nil, data: nil)
        }
    }

    func disable() {
        isEnabled.setFalseIfTrue {
            logger.log(level: .info, message: "LogEventCollector disabled.", error: nil, data: nil)
        }
    }

    func trackLog(body: String, severity: LogSeverity, attributes: [String: AttributeValue]) {
        guard isEnabled.get() else { return }
        guard !body.isEmpty else {
            logger.log(level: .warning, message: "LogEventCollector: Log message is empty. This event will be dropped.", error: nil, data: nil)
            return
        }
        guard severity.severityNumber >= configProvider.logMinSeverity else { return }
        guard !configProvider.shouldDiscardLog(body: body) else { return }
        guard attributeValueValidator.validateAttributes(name: "log", attributes: attributes) else { return }

        let data = LogData(severityText: severity.severityText,
                           severityNumber: severity.severityNumber,
                           body: String(body.prefix(Int(configProvider.maxLogBodyLength))))
        let userDefinedAttributes = EventSerializer.serializeUserDefinedAttribute(attributes)

        signalProcessor.trackUserTriggered(data: data,
                                           timestamp: timeProvider.now(),
                                           type: .log,
                                           attributes: nil,
                                           sessionId: nil,
                                           attachments: nil,
                                           userDefinedAttributes: userDefinedAttributes,
                                           threadName: nil,
                                           needsReporting: true)
    }
}
